### ISSUE TITLE
/oauth2/userinfo should work like a proxied endpoint

### ISSUE THREAD

**Issue by rassie** [oauth2-proxy/oauth2-proxy#985]
This might be something that could be split in two issues, but I think both can be solved with one patch.

## Expected Behavior

1. `/oauth2/userinfo` should return 401 if the token is expired
2. `/oauth2/userinfo` should include response headers if such injection is configured

## Current Behavior

1. Currently, `/oauth2/userinfo` returns HTTP code 401 if there is no authentication cookie, but always returns user information from the token even if this token is expired. I can't judge whether this is a security concern, but this behavior is not consistent.

2. Currently, `/oauth2/userinfo` is not injecting response headers as every proxied endpoint does if configured. I would like to argue that this makes some workflows more complicated than necessary. A simple example: if I need to fetch user's full name, I need to extract it from the ID token. There is no endpoint to fetch tokens, but it can be injected into response headers with

```
injectResponseHeaders:
  - name: Authorization
    values:
      - claim: id_token
        prefix: "Bearer "
```

However, to actually fetch the token I'd need to make a request to some unrelated custom endpoint in the backend (e.g. a healthcheck), which is unintuitive at best. 

## Possible Solution

Both issues could be solved if `/oauth2/userinfo` would be handled more like a proxied request. 

1. Every proxied request validates tokens, so with expired token, `/oauth2/userinfo` would return 401 as expected
2. Every proxied request contains injected response headers, if defined, so in combination with the first issue, a simple request to `/oauth2/userinfo` could provide a frontend with all the information it needs about current authenticated situation.

**Comment by NickMeves** [comment]
I think I can address (1)

Everything in our system for `userinfo` is based on our concept of `SessionState` which is built up during the initial OAuth flow & a call to `EnrichSession` which depending on provider might pull out session fields from token claims, or use an Access Token to hit a provider's userinfo endpoint to get the user details.

The reason we rely on our view of `User`, `Email` & `Groups` in our session for AuthZ logic & userinfo and disregard any data directly in ID Tokens claims at this point is due to widespread provider support. While OIDC providers always have that data in a token claim, pure OAuth2 Access Token only providers require that data to be fetched from their provider userinfo endpoint.

So we fetch it once during the initial authentication flow and store it in our concept of a session (so those providers wouldn't need to constantly hit those APIs to get user data on every proxied request).

Our session's lifespan is additionally its own thing, completely decoupled from Access/Refresh Token TTLs. So while we recommend you line the times up -- you could conceivably make the OAuth2 Proxy session as long or as short as your want regardless of the underlying token `exp`.

**Comment by NickMeves** [comment]
For (2) -- I'll need to weigh this one.

`X-Forwarded-*` forwarded request headers (and the parallel `X-Auth-Request-*` headers that are response headers intended for use with nginx subrequest or kubernetes ingress architectures) are the means to pass authenticated user session data to the upstream endpoints (for them to use as they wish with their AuthZ and RBAC logic).

All the `/oauth2/*` direct endpoints don't use them since they aren't intended to be in the request chain to an upstream endpoint.

Do you have more details on the architecture you are thinking where you need `/oauth2/userinfo` details in a response header in addition to the JSON payload that already has them?

**Comment by JoelSpeed** [comment]
Just wanted to add that for 2, the behaviour you're describing is somewhat similar to our current `/oauth2/auth` endpoint. It will inject headers that you have configured if the request is successful and return 401 if not.

In terms of the user info endpoint, it hasn't been touched in a while. We are making some big architecture changes within the codebase and this may have been overlooked. The way I expected this page to work is similar to that which you described, 401 with not details returned if not logged in, 200 with up to date information from the session if you are logged in. Sounds like this could be a bug 🤔

**Comment by rassie** [comment]
Oops. This is of course still valid. @NickMeves please re-open, thanks!

**Comment by JoelSpeed** [comment]
Still valid, and I think, may be fixed by #1128
